app/src/main/java/com/oracle/javafx/scenebuilder/app/welcomedialog/WelcomeDialogWindowController.java

```diff
@@ -141,17 +171,97 @@ public static WelcomeDialogWindowController getInstance() {
-        sceneBuilderApp.handleOpenFilesAction(Arrays.asList(projectPath));
-        getStage().hide();
+        handleOpen(List.of(projectPath));
     }
 
     @FXML
     private void openDocument() {
-        // Right now there is only one window open by default
-        DocumentWindowController documentWC = sceneBuilderApp.getDocumentWindowControllers().get(0);
-        sceneBuilderApp.performControlAction(SceneBuilderApp.ApplicationControlAction.OPEN_FILE, documentWC);
-        getStage().hide();
+        FileChooser fileChooser = new FileChooser();
+        fileChooser.getExtensionFilters().add(
+                new FileChooser.ExtensionFilter(I18N.getString("file.filter.label.fxml"), "*.fxml")
+        );
+        fileChooser.setInitialDirectory(EditorController.getNextInitialDirectory());
+
+        List<File> fxmlFiles = fileChooser.showOpenMultipleDialog(getStage());
+
+        // no file was selected, so nothing to do
+        if (fxmlFiles == null)
+            return;
+
+        List<String> paths = fxmlFiles
+                .stream()
+                .map(File::toString)
+                .collect(Collectors.toList());
+
+        handleOpen(paths);
+    }
+
+    private void handleOpen(List<String> paths) {
+        if (sceneBuilderApp.startupTasksFinishedProperty().get()) {
+            sceneBuilderApp.handleOpenFilesAction(paths);
+            getStage().hide();
+        } else {
+            showMasker(() -> {
+                sceneBuilderApp.handleOpenFilesAction(paths);
+                getStage().hide();
+            });
+        }
+    }
+
+    private void showMasker(Runnable onEndAction) {
+        // swap roots to allow adding masker if running for the first time
+        if (newRoot == null) {
+            initMasker();
+        }
+
+        rootPane.setDisable(true);
+        newRoot.getChildren().add(masker);
+
+        sceneBuilderApp.startupTasksFinishedProperty().addListener((o, old, isFinished) -> {
+            if (isFinished) {
+                Platform.runLater(() -> {
+                    onEndAction.run();
+
+                    // restore root in case welcome dialog is opened again
+                    rootPane.setDisable(false);
+                    newRoot.getChildren().remove(masker);
+                });
+            }
+        });
+    }
+
+    private void initMasker() {
+        newRoot = new StackPane();
+        newRoot.getStylesheets().setAll(rootPane.getStylesheets());
+
+        var label = new Label();
+        label.getStyleClass().add("progress-label");
+        label.setText(I18N.getString("welcome.loading.label"));
+
+        var progress = new ProgressIndicator();
+        progress.getStyleClass().add("progress");
+
+        masker = new VBox(progress, label);
+        masker.setAlignment(Pos.CENTER);
+
+        getScene().setRoot(newRoot);
+
+        rootPane.prefWidthProperty().bind(getScene().widthProperty());
+        rootPane.prefHeightProperty().bind(getScene().heightProperty());

```

Comment:
I don't understand why you do a binding here for `rootPane`. Probably better to do it for `newRoot`?  Also, when masker is removed, shouldn't you unbind?

Reply:
`rootPane` is BorderPane (which comes from fxml). Since it is no longer root (`StackPane newRoot` is when the masker is added), it does not grow to fill the parent (the StackPane). The above binding is the easiest way I found to make `rootPane` (and therefore `newRoot`) to fill the parent.

This is not related to masker per se, so no unbinding is needed.